cranelift-sizeless-vector.md

```diff
@@ -0,0 +1,46 @@
+# Summary
+
+This RFC proposes a way to handle flexible vector types as specified at: https://github.com/WebAssembly/flexible-vectors
+
+[summary]: #summary
+
+The proposal is to introduce new sizeless vector types into Cranelift, that:
+- Express a vector, with a lane type and size, but a target-defined number of lanes.
+- Are denoted by prefixing the lane type with 'sv' (sizeless vector): svi16, svi32, svf32, etc...
+- Have a minimum width of 128-bits, meaning we can simply map to existing simd-128 implementations.
```

Comment:
Riscv allows smaller vectors, right?

Reply:
Yes, apparently so, but the flexible vector spec does not.

Reply:
Also, the flexible vectors specification might introduce the requirement that the underlying vector length is a power of 2, even though the Arm SVE and the RISC-V V extension, I suppose, do not have similar restrictions. Whether our IR should cover more than the common use cases (such as the Wasm flexible vectors, power-of-2 vector sizes, etc.) is something to discuss.

Reply:
It seems to me that supporting more general expressivity at the CLIF level is a good thing to establish early on if we plan to do it at all, or else we'll bake in a lot of assumptions to code that operates on these types, over time. So non-power-of-2 sizes is something to include (and test for) early. It does seem valuable to me to support this if underlying hardware commonly does.

On the other hand, we don't want to come up with a design that requires handling difficult edge-cases; emulating smaller vector support than the fundamental unit seems like a good example of that. (Edge-cases if one were to try to emulate with bigger ops involve e.g. ensuring loads/stores don't run beyond the end of the heap, and having smaller alignments than usual.) So scaling down to 64 bits seems like an unreasonable goal just because one possible future platform supports it. So, I'd tentatively suggest a 128-bit minimum size.